### PR TITLE
Allow policies to be omitted from ScubaGear

### PR DESCRIPTION
**Pull request by adhilto**
## 🗣 Description ##
Make it so that policies can be omitted from the ScubaGear reports.

## 💭 Motivation and context ##
Closes https://github.com/cisagov/ScubaGear/issues/740.
Closes https://github.com/cisagov/ScubaGear/issues/739.
Closes https://github.com/cisagov/ScubaGear/issues/738.

## 🧪 Testing ##
I'd recommend that the reviewers start by reading the documentation I added for this feature, to get a good overview of the approach taken: https://github.com/cisagov/ScubaGear/blob/739-allow-any-given-policy-to-be-ignored/docs/configuration/configuration.md#omit-policies.

Testing I've done:
- Added 8 new Pester cases to cover these changes
- Manually tested without a config file to ensure normal executions is unchanged
- Manually tested excluding several policies without expiration dates
- Manually tested excluding policies without specifying the rationale
- Manually tested excluding policies with expirations date, both in the past and future
- Manually tested excluding policies with a malformed expiration date
- Manually tested generating a config file with policy omissions via New-Config, including some edge cases, such as malformed policy ids, variations in capitalization, and unexpected product names in the policy IDs.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] Changes are sized such that they do not touch excessive number of files.
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant repo and/or project documentation updated to reflect these changes.
- [x] Unit tests added/updated to cover PowerShell and Rego changes.
- [x] Functional tests added/updated to cover PowerShell and Rego changes.
- [x] All relevant functional tests passed.
- [x] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [ ] PR passed smoke test check.
- [ ] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [ ] Resolved all merge conflicts on branch
- [ ] Notified merge coordinator that PR is ready for merge via comment mention

## ✅ Post-merge checklist ##

- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

File: PowerShell/ScubaGear/Modules/Orchestrator.psm1
Comment (gdasher) [inline PowerShell/ScubaGear/Modules/Orchestrator.psm1:1]: It looks like as implemented this won't change the Json version of the report.  I think that omissions should be included in the json output as well, that way for example we can build graphs to track them historically in MTA without rejoining with the configuration state (which may have changed from the moment of assessment to present).

Comment (adhilto) [reply PowerShell/ScubaGear/Modules/Orchestrator.psm1:1]: Great callout. It does* (I just double checked to be safe).

The asterisk: it changes the ScubaResults.json file, but not the Rego output, TestResults.json, which is probably what you've picked up on.

ScubaResults.json is the json equivalent of the HTML report and is the file used for MTA. Here's an example of what it looks like when run with an omission:
image

TestResults.json is just an intermediary file (arguably with a misleading name) and isn't used for MTA.

Comment (gdasher) [reply PowerShell/ScubaGear/Modules/Orchestrator.psm1:1]: Ahh I see now how the new version works.  Ok, thanks for checking.  Looks good!

**Comment by tkol2022** [comment]
### Feedback on example config file

The example MS.TEAMS.6.1v1 provided in the sample config file is confusing with respect to the sample rationale. If the service is provided by M365 Defender and the Defender baseline checks for the implementation of this policy, then what is the purpose of having this policy in Teams to begin with? I believe this has a good chance of confusing users.
Instead I recommend you provide an example rationale such as "The DLP capability required for Teams is implemented by third party product X which ScubaGear does not have the ability to check" or something like that.

I also recommend that you trim the number of example policy IDs that you have. I think by the time the user gets to policy 7.1 in the example config they will get the point so you could get rid of examples 7.2, 8.1, 8.2 and we wouldn't lose anything.

**Comment by adhilto** [comment]
> ### Feedback on example config file
> The example MS.TEAMS.6.1v1 provided in the sample config file is confusing with respect to the sample rationale. If the service is provided by M365 Defender and the Defender baseline checks for the implementation of this policy, then what is the purpose of having this policy in Teams to begin with? I believe this has a good chance of confusing users. Instead I recommend you provide an example rationale such as "The DLP capability required for Teams is implemented by third party product X which ScubaGear does not have the ability to check" or something like that.
> 
> I also recommend that you trim the number of example policy IDs that you have. I think by the time the user gets to policy 7.1 in the example config they will get the point so you could get rid of examples 7.2, 8.1, 8.2 and we wouldn't lose anything.

Done

**Comment by tkol2022** [comment]
I made a few tweaks to the configuration.md because i felt the term "excluded" could get conflated with AAD conditional access policy exclusions. we should stick with the term "omitted" and omissions.

**Comment by tkol2022** [comment]
In configuration.md, I think we can delete this paragraph. It is confusing and I don't think it adds anything. Do you think anyone is likely to send policy ids as command line arguments to New-Config? It is easier to just modify the yaml config file in my opinion.

"Policy omissions can be provided to the `New-Config` function as a comma-separated list of policy IDs under the `OmitPolicy` parameter. However, the `New-Config` function does not allow you to specify the rationales or expiration dates via the commandline; these must be manually entered into the resulting config file."

**Comment by tkol2022** [comment]
The expiration date feature is a nice touch!

**Comment by tkol2022** [comment]
Will the following line perform a case-insensitive match? If not, should we make is so that case does not matter?

`if ($Config.OmitPolicy.psobject.properties.name -Contains $ControlId) `

**Comment by tkol2022** [comment]
What are your thoughts on changing the following?
Test disabled by user.
to
Test omitted by user.

**Comment by tkol2022** [comment]
### Bug
If the user accidentally provides two Rationale fields ScubaGear crashes. The same problem occurs if you have more than one omit policy entry for the same policy id.

image
image

image

**Comment by tkol2022** [comment]
> Will the following line perform a case-insensitive match? If not, should we make is so that case does not matter?
> 
> `if ($Config.OmitPolicy.psobject.properties.name -Contains $ControlId) `

I tested this. It ignores case which is great for flexibility. No change needed.
image

**Comment by tkol2022** [comment]
### Test summary

- Tested prior date, today and future date for the Expiration field and the tool behaved as expected.
- Tested single quote in Rationale field and tool behaved as expected.
- Tested rationale missing and tool behaved as expected.
- Tested malformed date and the tool behaved as expected.
- Tested without rationale and without date and the tool behaved as expected.
- The only test that produced a bug was when you have more than one of the same policy or you have multiple rationales. See separate comment with details.

**Review by tkol2022 — APPROVED** [review]
I found one bug and left some comments for a couple of other tweaks. If you can address those I think we are in good shape. Nice work. Overall the system behaved as expected under various conditions. Also I asked a related question on Slack which was answered by the team about how the downstream system interprets the outputs from this new functionality.

File: PowerShell/ScubaGear/Modules/Support/Support.psm1 [hunk 1]
```diff
@@ -849,6 +863,46 @@ function New-Config {
...
 
+    $OmissionNamespace = "OmitPolicy"
+
+    # List to track which policies the user specified in $OmitPolicies are properly formatted
+    $OmitPolicyValidated = @()
+
+    # Hashmap to structure the ignored policies template
+    $config[$OmissionNamespace] = @{}
+
+    foreach ($Policy in $OmitPolicy) {
```
Comment (mitchelbaker-cisa) [inline PowerShell/ScubaGear/Modules/Support/Support.psm1:874]: This logic validates the formatting of $OmitPolicies when a user first invokes New-Config, but I'm not seeing any validation in cases where the config.yml is modified after creation. If I incorrectly set a key in `OmitPolicy` to `MSAAD4.1v1` or `MS.GWS.AAD4.2v1` these omissions are skipped and the policy runs normally. 

Perhaps an additional unit test to verify malformed policy IDs are excluded from CreateReport.psm1 and the policy is expected to run by default?

**Review by mitchelbaker-cisa — COMMENTED** [review]
Testing:

Ran ScubaGear without a config file, runs as expected. 
Ran ScubaGear with a config file:
- with valid policy ID, rationale, expiration
- with invalid policy ID
- with null Expiration value (results in skipped omission for given policy)
- with non-string Expiration value (results in skipped omission for given policy)

Verified "OmitPolicy" is included in ProviderSettingsExport 
Verified Result: "Omitted" is in individual report json and in combined json with -mergeJson flag

Left one comment to consider but otherwise looks good.

**Comment by adhilto** [comment]
> In configuration.md, I think we can delete this paragraph. It is confusing and I don't think it adds anything. Do you think anyone is likely to send policy ids as command line arguments to New-Config? It is easier to just modify the yaml config file in my opinion.
> 
> "Policy omissions can be provided to the `New-Config` function as a comma-separated list of policy IDs under the `OmitPolicy` parameter. However, the `New-Config` function does not allow you to specify the rationales or expiration dates via the commandline; these must be manually entered into the resulting config file."

I agree that it's probably easier to reference the example config and modify their config file manually. I'm ok with deleting the paragraph.

**Comment by adhilto** [comment]
> ### Bug
> If the user accidentally provides two Rationale fields ScubaGear crashes. The same problem occurs if you have more than one omit policy entry for the same policy id.

I think this bug is out of scope for this PR, I recommend opening a separate issue for it. This bug is in the orchestrator and/or the ScubaConfig.psm1 `LoadConfig` function. The YAML specification actually does not allow duplicate keys, so in your test, you've engineered invalid yaml. The problem this surfaces is that the orchestrator does not catch that invalid yaml was provided. The first error message in your screenshot is a bit cryptic, but the fix might be to wrap this line (in ScubaConfig.psm1) in a try/catch:
image

Additionally, for whatever reason, this `if` in Orchestraror.psm1 failed to detect that the config failed to load.
image

**Comment by tkol2022** [comment]
> Additionally, for whatever reason, this `if` in Orchestraror.psm1 failed to detect that the config failed to load. image

I opened an issue. BTW, from what I can tell, the reason that code in Orchestrator does not work as expected is because when the LoadConfig function produces an error, Orchestrator will not execute the if statement you mentioned. Instead it will go to the nearest catch () block, right?

**Comment by adhilto** [comment]
> > Additionally, for whatever reason, this `if` in Orchestraror.psm1 failed to detect that the config failed to load. image
> 
> I opened an issue. BTW, from what I can tell, the reason that code in Orchestrator does not work as expected is because when the LoadConfig function produces an error, Orchestrator will not execute the if statement you mentioned. Instead it will go to the nearest catch () block, right?

Honestly, I'm not really sure what will happen. Unless I'm just not looking deep enough, this block of code isn't within a `try` at all. Definitely more digging needed here.

File: PowerShell/ScubaGear/Modules/Support/Support.psm1 (hunk 1 above)
Comment (adhilto) [reply PowerShell/ScubaGear/Modules/Support/Support.psm1:874]: Done. I also modified the load config function to output warnings for invalid IDs. Tagging @tkol2022 for awareness.
image

**Comment by adhilto** [comment]
> What are your thoughts on changing the following? Test disabled by user. to Test omitted by user.

Changed

**Review by mitchelbaker-cisa — APPROVED** [review]

**Comment by adhilto** [comment]
@nanda-katikaneni Looks like this one is good to go

**Pull request merged**